docs/proposals/message-proto-design.md

```diff
@@ -0,0 +1,346 @@
+
+## Need to do
+
+*  Change the model message to protobuf message  directly
+*  Solve the filed index problem of `Resource`
+*  Consider compatibility  
+
+## How to do 
+
+##  Change the model message to protobuf message  directly
+
+create proto file in `model` dir to define message struct. and use `go generate`(will move to makefile) to create message.pb.go file 
+
+```
+syntax = "proto3";
+package model;
+import "google/protobuf/any.proto";
+
+message MessageHeader {
+    string ID = 1;
+    string ParentID = 2;
+    int64 Timestamp = 3;
+    bool Sync = 4;
+}
+message MessageRoute {
+    string Source = 1;
+    string Group = 2;
+    string Operation = 3;
+    google.protobuf.Any Resource = 4;
+}
+
+message Message {
+    MessageHeader Header = 1;
+    MessageRoute Router = 2;
+    google.protobuf.Any Data = 3;
+}
+```
+
+we use `Any type` to define `Resource` and `Data` field.  `Data` field is the original `Content` field.
+
+##  Solve the filed index problem of `Resource`
+
+this problem is the most important, currently, beehive define path string to stand for resource, such as `/node/{nodeID}/{namespace}/{resourceType}/{resource}`, different path stand for different resource.  but we need hardcord to parse `nodeid` ,`resourceType`,`resourceid`. It's not so extensible, so we need a more extended form. we need change `Resource` to golang `interface{}`. but in protobuf, there is not `interface` define. we need use `Any` type.
+
+```
+message MessageRoute {
+    string Source = 1;
+    string Group = 2;
+    string Operation = 3;
+    google.protobuf.Any Resource = 4;
+}
+```
+we use `github.com/gogo/protobuf`  package not offical `github.com/golang/protobuf` package . 
+
+`gogo/protobuf/types` package provides function to encodes `Any` types 
+```
+	func MarshalAny(pb proto.Message) (*Any, error) 
+	func UnmarshalAny(any *Any, pb proto.Message) error 
+	
+```
+but they can only support convert `proto.Message` into `Any` type, or convert `Any` type into `proto.Message`.  they can not convert custom type into `Any` type.
+
+so We need to support convert custom type into `Any` types, and convert `Any` types into custom types.
+
+in `pkg/core/typeurl` package of `github.com/kubeedge/beehive` project, we support custom types conversion to Any types. there are some functions:
+
+* `func RegisterType(v interface{}, args ...string)` , this function Register custom types . if we want string types can conversion to Any types. we need to add 
+ ```
+ RegisterType("", "string.content")
+ ``` 
+if we want bool types can conversion to Any types. we need to add
+```
+RegisterType(true, "bool.content")
+```
+
+if we want v1.Pod types ("k8s.io/api/core/v1") can conversion to Any types. we need to add
+``` 
+RegisterType(v1.Pod{}, "v1.pod.content")
+```
+the second parameter used to generate TypeUrl of `Any` type. the TypeUrl must be unique, otherwise it will panic.
```

Comment:
what is TypeUrl ?

Reply:
TypeUrl is A URL/resource name that uniquely identifies the type of the serialized protocol buffer message.

https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto